lib/client/client.go

```diff
@@ -1997,60 +1997,81 @@ func acceptWithContext(ctx context.Context, l net.Listener) (net.Conn, error) {
 
 // listenAndForward listens on a given socket and forwards all incoming
 // commands to the remote address through the SSH tunnel.
-func (c *NodeClient) listenAndForward(ctx context.Context, ln net.Listener, remoteAddr string) {
+func (c *NodeClient) listenAndForward(ctx context.Context, ln net.Listener, localAddr string, remoteAddr string) {
 	defer ln.Close()
-	defer c.Close()
 
-	for {
+	log := log.WithField("localAddr", localAddr).WithField("remoteAddr", remoteAddr)
+
+	log.Infof("Starting port forwarding")
+
+	for ctx.Err() == nil {
 		// Accept connections from the client.
 		conn, err := acceptWithContext(ctx, ln)
 		if err != nil {
-			log.Errorf("Port forwarding failed: %v.", err)
-			break
+			if ctx.Err() == nil {
+				log.WithError(err).Errorf("Port forwarding failed.")
+			}
+			continue
 		}
 
 		// Proxy the connection to the remote address.
 		go func() {
-			err := proxyConnection(ctx, conn, remoteAddr, c.Client)
-			if err != nil {
-				log.Warnf("Failed to proxy connection: %v.", err)
+			var err error
+			if err = proxyConnection(ctx, conn, remoteAddr, c.Client); err != nil {
+				log.WithError(err).Warnf("Failed to proxy connection.")
 			}
```

Comment:
This was conscious rewrite: `err` here is shadowing `err` in parent scope. This is what want, but I want to avoid a possible error when someone "fixes" the code by rewriting `err := ...` as `err = ...`. With explicit `var` declaration this is much harder mistake to make.

Reply:
Yeah, so it was correct before I switched to `var err error` - we had `:=` instead of `=`, but this demonstrates my point: it is easy to miss such issues. But yeah, `var err error` looks weird, I'll find a better way to express it.

Reply:
I had `var err error` in two separate functions, I missed one of those when replacing it with `:=` as per feedback. Should be good now.